File: data-access-layer/modules/reporting/svc.ts
import { pool } from "../../db/db.ts";
import type { ParentDocument } from "./model.ts";
import {
  type IReportingRepo,
  type IS3Repository,
  ReportingRepo,
  S3Repository,
} from "./repo.ts";
import type { DocumentType } from "../../shared/types/reporting.ts";
import { redisClient } from "../../db/redis-client.ts";
import { extname } from "path";

export interface IReportingSvc {
  getParentDocumentByType(
    userId: string,
    documentType: DocumentType,
  ): Promise<{ data?: ParentDocument; error?: Error }>;
  getParentDocumentTypeByDocumentId(
    parentDocumentId: string,
  ): Promise<{ data?: DocumentType; error?: Error }>;
  saveParentDocument(
    userId: string,
    documentType: DocumentType,
    file: File,
  ): Promise<{ data?: ParentDocument; error?: Error }>;
  getDocumentURLByFilePath(
    bucket: string,
    key: string,
    expiresIn: number,
  ): Promise<{ data?: string; error?: Error }>;
  getParentDocumentURLByDocumentID(
    docId: string,
  ): Promise<{ data?: string; error?: Error }>;
}

class ReportingSvc implements IReportingSvc {
  private repo: IReportingRepo;
  private s3Repo: IS3Repository;
  constructor(repo?: IReportingRepo, s3Repo?: IS3Repository) {
    this.repo = repo ?? new ReportingRepo(pool, redisClient);
    this.s3Repo = s3Repo ?? new S3Repository();
  }

  async getParentDocumentByType(
    userId: string,
    documentType: DocumentType,
  ): Promise<{ data?: ParentDocument; error?: Error }> {
    return await this.repo.getParentDocumentByType(userId, documentType);
  }

  async getParentDocumentTypeByDocumentId(
    parentDocumentId: string,
  ): Promise<{ data?: DocumentType; error?: Error }> {
    return await this.repo.getParentDocumentTypeByDocumentId(parentDocumentId);
  }

  async saveParentDocument(
    userId: string,
    documentType: DocumentType,
    file: File,
  ): Promise<{ data?: ParentDocument; error?: Error }> {
    const filePath = `parents/${userId}/documents/${documentType}${extname(file.name)}`;
    const { error } = await this.s3Repo.uploadFile("mybucket", filePath, file);
    if (error) return { data: undefined, error };
    const result = await this.repo.saveParentDocument(
      userId,
      documentType,
      filePath,
    );
    if (result.error) return { data: undefined, error: result.error };
    return result;
  }

  async getDocumentURLByFilePath(
    bucket: string,
    key: string,
    expiresIn: number = 3600,
  ): Promise<{ data?: string; error?: Error }> {
    return await this.s3Repo.getDocumentURLByFilePath(bucket, key, expiresIn);
  }

  async getParentDocumentURLByDocumentID(
    docId: string,
  ): Promise<{ data?: string; error?: Error }> {
    const { data, error } =
      await this.repo.getParentDocumentFilePathByDocumentID(docId);
    if (error) return { data: undefined, error };
    return await this.s3Repo.getDocumentURLByFilePath(data);
  }
}

const reportingSvc = new ReportingSvc();
export default reportingSvc;

File: apps/backend/src/modules/reporting/svc.py
import os
from abc import ABC, abstractmethod
from uuid import UUID
from fastapi import UploadFile, File
from src.modules.reporting.repo import IReportingRepo
from src.modules.reporting.s3 import IS3Repository
from src.shared.types.modules.reporting.enum import CHILD_DOCUMENT_TYPE, DOCUMENT_TYPE
from src.shared.types.modules.reporting.model import (
    ChildDocument,
    ParentDocument,
    ParentPartnerDocument,
)


class IReportingSvc(ABC):
    @abstractmethod
    async def get_parent_document_by_type(
        self,
        user_id: UUID,
        document_type: DOCUMENT_TYPE,
    ) -> ParentDocument:
        pass

    @abstractmethod
    async def get_parent_document_url_by_document_id(self, doc_id: UUID) -> str:
        pass

    @abstractmethod
    async def get_document_url_by_file_path(
        self,
        path: str,
    ) -> str:
        pass

    @abstractmethod
    async def save_parent_document(
        self,
        user_id: UUID,
        document_type: DOCUMENT_TYPE,
        file: UploadFile = File(...),
    ) -> ParentDocument:
        pass

    @abstractmethod
    async def get_child_document_by_type(
        self,
        child_id: UUID,
        document_type: CHILD_DOCUMENT_TYPE,
    ) -> ChildDocument:
        pass

    @abstractmethod
    async def save_child_document(
        self,
        child_id: UUID,
        document_type: CHILD_DOCUMENT_TYPE,
        file: UploadFile = File(...),
    ) -> ChildDocument:
        pass

    @abstractmethod
    async def get_child_document_url_by_document_id(self, doc_id: UUID) -> str:
        pass

    @abstractmethod
    async def get_parent_partner_document_by_type(
        self,
        partner_id: UUID,
        document_type: DOCUMENT_TYPE,
    ) -> ParentPartnerDocument:
        pass

    @abstractmethod
    async def save_parent_partner_document(
        self,
        partner_id: UUID,
        document_type: DOCUMENT_TYPE,
        file: UploadFile = File(...),
    ) -> ParentPartnerDocument:
        pass


class ReportingSvc(IReportingSvc):
    def __init__(self, repo: IReportingRepo, s3_repo: IS3Repository) -> None:
        self.repo = repo
        self.s3_repo = s3_repo

    async def get_parent_document_by_type(
        self,
        user_id: UUID,
        document_type: DOCUMENT_TYPE,
    ) -> ParentDocument:
        return await self.repo.get_parent_document_by_type(
            user_id=user_id, document_type=document_type
        )

    async def get_parent_document_url_by_document_id(self, doc_id: UUID) -> str:
        path = await self.repo.get_parent_document_file_path_by_document_id(
            doc_id=doc_id
        )
        return await self.get_document_url_by_file_path(path=path)

    async def get_document_url_by_file_path(
        self,
        path: str,
    ) -> str:
        return await self.s3_repo.get_document_url_by_file_path(key=path)

    async def save_parent_document(
        self,
        user_id: UUID,
        document_type: DOCUMENT_TYPE,
        file: UploadFile = File(...),
    ) -> ParentDocument:
        _, ext = os.path.splitext(file.filename)
        file_path = f"parents/{user_id}/documents/{document_type}{ext}"
        file_content = await file.read()
        await self.s3_repo.upload_file(key=file_path, file_content=file_content)
        return await self.repo.save_parent_document(
            user_id=user_id, document_type=document_type, file_path=file_path
        )

    async def get_child_document_by_type(
        self,
        child_id: UUID,
        document_type: CHILD_DOCUMENT_TYPE,
    ) -> ChildDocument:
        return await self.repo.get_child_document_by_type(
            child_id=child_id, document_type=document_type
        )

    async def save_child_document(
        self,
        child_id: UUID,
        document_type: CHILD_DOCUMENT_TYPE,
        file: UploadFile = File(...),
    ) -> ChildDocument:
        _, ext = os.path.splitext(file.filename)
        file_path = f"children/{child_id}/documents/{document_type}{ext}"
        file_content = await file.read()
        await self.s3_repo.upload_file(key=file_path, file_content=file_content)
        return await self.repo.save_child_document(
            child_id=child_id, document_type=document_type, file_path=file_path
        )

    async def get_child_document_url_by_document_id(self, doc_id: UUID) -> str:
        path = await self.repo.get_child_document_file_path_by_document_id(
            doc_id=doc_id
        )
        return await self.get_document_url_by_file_path(path=path)

    async def get_parent_partner_document_by_type(
        self,
        partner_id: UUID,
        document_type: DOCUMENT_TYPE,
    ) -> ParentPartnerDocument:
        return await self.repo.get_parent_partner_document_by_type(
            partner_id=partner_id, document_type=document_type
        )

    async def save_parent_partner_document(
        self,
        partner_id: UUID,
        document_type: DOCUMENT_TYPE,
        file: UploadFile = File(...),
    ) -> ParentPartnerDocument:
        _, ext = os.path.splitext(file.filename)
        file_path = f"parents/partners/{partner_id}/documents/{document_type}{ext}"
        file_content = await file.read()
        await self.s3_repo.upload_file(key=file_path, file_content=file_content)
        return await self.repo.save_parent_partner_document(
            partner_id=partner_id, document_type=document_type, file_path=file_path
        )
